### ISSUE TITLE
ICE: type mismatch comparing repeat and repeat

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#45044]
## STR
```
$ rustc --version
rustc 1.22.0-nightly (185cc5f26 2017-10-02)
```

```Rust
const X: [u8; 1] = [0; 1];

fn main() {
    match &X {
        &X => println!("a"),
        _ => println!("b"),
    };
}
```

## Expected Result

code compiles and runs

## Actual Result
```
warning: unreachable pattern
 --> x.rs:6:9
  |
6 |         &X => println!("a"),
  |         ^^
  |
  = note: #[warn(unreachable_patterns)] on by default

error: internal compiler error: type mismatch comparing Aggregate(Repeat(Const { ty: u8, val: Integral(U8(0)) }, 1)) and Aggregate(Repeat(Const { ty: u8, val: Integral(U8(0)) }, 1))
 --> x.rs:6:10
  |
6 |         &X => println!("a"),
  |          ^

error: aborting due to previous error

```

**Comment by ysiraichi** [comment]
I think there is no one working on this issue yet, so can I work on this?
@nikomatsakis can you help me (it's my first time contributing to rust)?

**Comment by nikomatsakis** [comment]
@ysiraichi I agree nobody is working on this issue yet. =) I can try to leave a few tips. Maybe you can take some of these steps first.

It'd be good to know where the type error arises, first off. I'd run the compiler with `RUST_BACKTRACE=1`.

Second, the "unreachable pattern" message seems a bit strange. This no doubt arises -- in part -- because of the way that we handle constants in patterns. In short, the `&X` pattern will be internally expanded to `&[0; 1]` -- that is, the definition of the array is inlined. It's been a while since I looked at this code -- but I think that takes place in [`_match.rs`](https://github.com/rust-lang/rust/blob/a0dcecff90c45ad5d4eb60859e22bb3f1b03842a/src/librustc_const_eval/_match.rs) somewhere?

cc @oli-obk -- have you been looking at this code lately? have any tips?

**Comment by oli-obk** [comment]
Yes I have touched this code very recently, it will be fixed by #46882
